compositor_render/src/transformations/web_renderer.rs

```diff
@@ -85,25 +107,77 @@ impl WebRenderer {
         buffers: &[Arc<wgpu::Buffer>],
         target: &mut NodeTexture,
     ) -> Result<(), RenderWebsiteError> {
-        let mut controller = self.controller.lock().unwrap();
-        controller.send_sources(ctx, node_id.clone(), sources, buffers)?;
-
-        if let Some(frame) = controller.retrieve_frame() {
-            let target = target.ensure_size(ctx.wgpu_ctx, self.params.resolution);
-
-            self.bgra_texture.upload(ctx.wgpu_ctx, &frame);
-            self.bgra_to_rgba.convert(
-                ctx.wgpu_ctx,
-                (&self.bgra_texture, &self.bgra_bind_group),
-                target.rgba_texture(),
+        self.embedding_helper
+            .prepare_embedding(node_id, sources, buffers)
+            .map_err(|err| RenderWebsiteError::EmbeddingFailed(self.spec.url.clone(), err))?;
+
+        if let Some(frame) = self.retrieve_frame() {
+            let target = target.ensure_size(ctx.wgpu_ctx, self.spec.resolution);
+            self.website_texture.upload(ctx.wgpu_ctx, &frame);
+
+            let (textures, textures_info) = self.prepare_textures(sources);
+            let mut shader_params = self.shader_params.lock().unwrap();
+            shader_params.update(textures_info, ctx.wgpu_ctx);
+
+            self.render_website_shader.render_with_textures(
+                &shader_params.bind_group,
+                &textures,
+                target,
+                Default::default(),
+                None,
             );
         }
 
         Ok(())
     }
 
+    fn prepare_textures<'a>(
+        &'a self,
+        sources: &'a [(&NodeId, &NodeTexture)],
+    ) -> (Vec<Option<&'a Texture>>, Bytes) {
+        let mut source_textures = sources_to_textures(sources);
+        let source_transforms = TextureInfo::sources(&self.source_transforms.lock().unwrap());
+
+        let mut textures = Vec::new();
+        match self.spec.embedding_method {
+            WebEmbeddingMethod::NativeEmbeddingOverContent => {
+                textures.push(Some(self.website_texture.texture()));
```

Comment:
it's not great, but I understand why it's this way, lets maybe land it this way, but add an issue to rework this in the future